jetstream_pt/ray_engine_master.py

```diff
@@ -0,0 +1,170 @@
+    _ = ray.get(all_outputs)
+    return None
+
+  # pylint: disable-next=all
+  def init_decode_state(
+      self,
+  ) -> DecodeState:
+    all_outputs = []
+    for worker in self.engine_workers:
+      output = worker.init_decode_state_ray.remote()
+      all_outputs.append(output)
+    _ = ray.get(all_outputs)
+    return None
+
+  def prefill(
+      self,
+      *,
+      params: Any,  # Weights
+      existing_prefix: Optional[Prefix] = None,
+      padded_tokens: np.ndarray,  # PrefillInputs[np.ndarray],
+      true_length: int,
+  ) -> Prefix:
+    all_outputs = []
+    for worker in self.engine_workers:
+      output = worker.prefill_ray.remote(
+          params=params,
+          existing_prefix=existing_prefix,
+          padded_tokens=padded_tokens,
+          true_length=true_length,
+      )
+      all_outputs.append(output)
+    _ = ray.get(all_outputs)
+
+  def insert(
+      self,
+      prefix: Prefix,
+      decode_state: DecodeState,
+      slot: int,
+  ) -> DecodeState:
+    all_outputs = []
+    for worker in self.engine_workers:
+      output = worker.insert_ray.remote(
+          prefix=prefix, decode_state=decode_state, slot=slot
+      )
+      all_outputs.append(output)
+    _ = ray.get(all_outputs)
+
+  def generate(
+      self, params: Any, decode_state: DecodeState
+  ) -> tuple[None, engine_api.ResultTokens]:
+    all_outputs = []
+    for worker in self.engine_workers:
+      output = worker.generate_ray.remote(
+          params=params, decode_state=decode_state
+      )
+      all_outputs.append(output)
+    state, result_tokens = ray.get(all_outputs)[0]
+    return state, result_tokens
+
+  # pylint: disable-next=all
+  def get_tokenizer(self) -> tokenizer_pb2.TokenizerParameters:
+    # pylint: disable-next=all
+    return tokenizer_pb2.TokenizerParameters(path=self.tokenizer_path)
+
+  @property
+  def max_concurrent_decodes(self) -> int:
+    return self.batch_size
+
+  @property
+  def samples_per_slot(self) -> int:
+    return 1
+
+  @property
+  def max_prefill_length(self) -> int:
+    return self.context_length
+
+  @property
+  def colocated_cpus(self) -> Union[list[engine_api.CpuDevices], None]:
+    return jax.devices("cpu")[0]
+
+  def get_prefix_destination_sharding(self) -> Prefix:
+    "No implementation"
+    return None
+
+  @property
+  def mesh(self):
+    "No implementation"
+    return None
+
+
+# pylint: disable-next=all
+def create_pytorch_engine_ray_master(
+    tokenizer_path: str,
+    ckpt_path: Optional[str] = None,
+    samples_per_slot: int = 1,
+    bf16_enable: bool = False,
+    param_size: str = "7b",
+    context_length: int = 1024,
+    batch_size: int = 1,
+    max_decode_length: int = 4096,
+    model_name="llama",
+    quantize_weights=False,
+    quantize_kv=False,
+    max_cache_length=1024,
+) -> PyTorchEngineRayMaster:
+
+  ray.init(ignore_reinit_error=True)
+  pod_name = tpu.get_current_pod_name()
+  num_hosts = tpu.get_current_pod_worker_count()
```

Comment:
maybe we can add an assertion check here that the pod_name is not `None` and `num_hosts >= 1` ?

`(None, 0)` would be the situation if we're not running on a TPU VM at all